Add UpsertMulti operation

kahuang opened the pull request.

maitesin commented:
Hi @kahuang,

Thanks for your contribution, but can you write some test that checks the functionality you just added?

kahuang commented:
I have added some tests, hopefully they succeed in the CI ;). I'm having trouble getting the test harness to work locally (something around authorization for the connection)

kahuang commented:
Ok the tests are passing now, the one failing seems unrelated to my change (and potentially flaky?)

maitesin commented:
Hi @kahuang,

Thanks for adding these test, but I think there is a test missing. Can you add a test for the case when the error returned form `c.writeOp(&op, true)` returns an error different from the key being duplicate?

eminano commented on session.go:
```diff
@@ -3124,6 +3123,41 @@ func (c *Collection) Upsert(selector interface{}, update interface{}) (info *Cha
 
+// Same as Upsert, but sets multi to True so it can update multiple documents, but upsert if none
+// exist
+func (c *Collection) UpsertMulti(selector interface{}, update interface{}) (info *ChangeInfo, err error) {
```
Since this is basically the same as ```Upsert```, what do you think about creating a private upsert method that takes the ```updateOp``` as parameter and handles the actual upsert, allowing to reuse that part of the code?
It will make the code more maintainable and flexible for possible future additions.

eminano commented on session_test.go:
```diff
@@ -662,6 +662,31 @@ func (s *S) TestUpdate(c *C) {
+	changeInfo, err := coll.UpsertMulti(M{"k": 1}, M{"$set": M{"n": 42}})
+	c.Assert(err, IsNil)
+	c.Assert(changeInfo.Matched, Equals, len(ns))
+	c.Assert(changeInfo.Updated, Equals, len(ns)-1)
```
Test looks good, but you probably want to check that the value set is the one you expect in addition to checking the number of documents updated/matched. Have a look at ```TestUpsert``` for an example.

eminano reviewed (changes requested):
Please target development branch instead of master.

Thanks!

kahuang commented:
@eminano definitely makes sense. I've refactored Upsert/UpsertMulti to use a shared private upsert() method. I've also updated the tests to include checking the values of the document we have just upserted.

kahuang committed:
Add UpsertMulti operation

eminano committed:
Merge branch 'development' into upsert_multi

eminano commented:
@kahuang Thanks for your contribution! :)

eminano committed:
Merge branch 'development' into upsert_multi
